algoliasearch-common/src/main/java/com/algolia/search/objects/Query.java

```diff
@@ -12,7 +12,7 @@
 @SuppressWarnings({"unused", "WeakerAccess"})
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonIgnoreProperties(ignoreUnknown = true)
-public class Query implements Serializable {
+public class Query<T extends Query<?>> implements Serializable {
```

Comment:
@aseure Maybe could you add a comment on that line? Just to explain what this trick is exactly doing. I think this is not trivial for everyone.

Reply:
Maybe a name a bit more expressive than T would be good too, like "Derived". And I don't know if it's already the case, but you should hide the constructor of Query so no one can build it directly (doesn't play well with CRTP)

Reply:
@Ant-hem Will add a comment for sure.

@nitnelave Sure, `T` is probably not the best name. Old habits die hard. Regarding the `Query` constructor, we need to keep it since both class can be instantiated in different contexts (actually, the `Query` class is the most used one). What makes you think it's a bad idea? With C++, I'd guess the instantiated would become more complicated with instances of the base class but with Java, it seems that instantiation remains the same.

Reply:
The problem is that if you use Query directly, you're using the raw type, which is basically a Java hack to let you write `List<A> a = new List<>();`. It's not really supposed to be used as a type. What I would do if you need to have a query class is move the CRTP and setters to an abstract class (e.g. QueryImpl) and have query just be an extension of that class with the right params.

Plus in terms of design, inheriting from a concrete class is usually bad practice.

Reply:
You're right. I'll add a common abstract class from which both `Query` and `ConsequenceParams` inherit. Thanks.

Reply:
Done in second commit.